test/4_test.py

```diff
@@ -0,0 +1,315 @@
+
+        room_creation = requests.post('http://localhost:8080/_matrix/client/r0/createRoom', json = {
+            'visibility': 'public',
+        }, headers = {
+            'Authorization': self.access_token
+        }).json()
+        print('Room creation %s' % room_creation)
+        self.room_id = room_creation['room_id']
+
+        print('Created room %s' % self.room_id)
+        pass
+
+    # 2. Starting actual test
+    def test(self):
+        quoted_room_id = urllib.parse.quote("%s" % self.room_id)
+        print("Testing with room %s, access token %s" % (quoted_room_id, self.access_token))
+
+        # A set of good events. They should be preserved in history.
+        good_events = {}
+        # A set of bad events. They should either be intercepted before they reach history or be redacted from history.
+        bad_events = {}
+
+        print('Sending raw text, it should pass')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/nu-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text without any link',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        good_events[response['event_id']] = "raw text"
+
+
+        print('Sending with good URLs, it should pass')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/gu-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text with a good link to good.example.com',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A text without a good link to <a href="http://good.example.com">good.example.com</a>',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        good_events[response['event_id']] = "good URL"
+
+        print('Sending text with bad url in body, it should be rejected')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/bb-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text with a hidden link to evil.example.com',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        if response.get('event_id', None) is None:
+            print('Message was rejected immediately')
+        else:
+            # Message may be redacted later
+            bad_events[response['event_id']] = "bad URL in body"
```

Comment:
I'm a bit confused by this -- are we unsure what will happen?

Reply:
As far as I can tell, it's not specified whether the message is redacted immediately or later.